src/group_theory/monoid_localization.lean

```diff
@@ -266,25 +277,348 @@ show _ * _ = _, by rw [mul_inv_left, mul_one]
+    of `comm_monoid`s, `k ∘ f` is a localization map for `M` at `S`. -/
+@[to_additive "If `f : M →+ N` is a localization map for a submonoid `S` and `k : N ≃+ P` is an isomorphism of `add_comm_monoid`s, `k ∘ f` is a localization map for `M` at `S`."]
+def mul_equiv_map (k : N ≃* P) :
+  localization_map S P :=
+{ to_fun := k.to_monoid_hom.comp f.1,
+  map_units := λ y, is_unit_comp f k.to_monoid_hom y,
+  surj := λ v, let ⟨z, hz⟩ := k.to_equiv.surjective v in
+    let ⟨x, hx⟩ := f.3 z in ⟨x, show v * k _ = k _, by rw [←hx, k.map_mul, ←hz]; refl⟩,
+  eq_iff_exists := λ x y, (k.to_equiv.apply_eq_iff_eq _ _).trans $ f.4 _ _ }
+
+/-- Given `comm_monoid`s `M, P` and submonoids `S ⊆ M, T ⊆ P`, if `f : M →* N` is a localization
+    map for `S` and `k : P ≃* M` is an isomorphism of `comm_monoid`s such that `k(T) = S`, `f ∘ k`
+    is a localization map for `T`. -/
+@[to_additive "Given `comm_monoid`s `M, P` and submonoids `S ⊆ M, T ⊆ P`, if `f : M →* N` is a localization map for `S` and `k : P ≃* M` is an isomorphism of `comm_monoid`s such that `k(T) = S`, `f ∘ k` is a localization map for `T`."]
+def mul_equiv_comap {k : P ≃* M} (H : k.to_monoid_hom.map T = S) :
+  localization_map T N :=
+let H' : k.to_monoid_hom.comap S = T :=
+  H ▸ (submonoid.ext' $ T.1.preimage_image_eq k.to_equiv.injective) in
+{ to_fun := f.1.comp k.to_monoid_hom,
+  map_units := λ y, let ⟨z, hz⟩ := f.2 ⟨k y, H ▸ set.mem_image_of_mem k y.2⟩ in ⟨z, hz⟩,
+  surj := λ z, let ⟨x, hx⟩ := f.3 z in let ⟨v, hv⟩ := k.to_equiv.surjective x.1 in
+    let ⟨w, hw⟩ := k.to_equiv.surjective x.2 in ⟨(v, ⟨w, H' ▸ show k w ∈ S, from hw.symm ▸ x.2.2⟩),
+    show z * f.1 (k.to_equiv w) = f.1 (k.to_equiv v), by erw [hv, hw, hx]; refl⟩,
+  eq_iff_exists := λ x y, show f.1 _ = f.1 _ ↔ _, by erw f.4;
+    exact ⟨λ ⟨c, hc⟩, let ⟨d, hd⟩ := k.to_equiv.surjective c in
+    ⟨⟨d, H' ▸ show k d ∈ S, from hd.symm ▸ c.2⟩, by erw [←hd, ←k.map_mul, ←k.map_mul] at hc;
+      exact k.to_equiv.injective hc⟩, λ ⟨c, hc⟩, ⟨⟨k c, H ▸ set.mem_image_of_mem k c.2⟩,
+    by erw ←k.map_mul; rw [hc, k.map_mul]; refl⟩⟩ }
+
+/-- Given localization maps `f : M →* N, k : P →* U` for submonoids `S, T` respectively, an
+    isomorphism `j : M ≃* P` such that `j(S) = T` induces an isomorphism of localizations
+    `N ≃* U`. -/
+@[to_additive "Given localization maps `f : M →+ N, k : P →+ U` for submonoids `S, T` respectively, an isomorphism `j : M ≃+ P` such that `j(S) = T` induces an isomorphism of localizations `N ≃+ U`."]
+noncomputable def mul_equiv_of_mul_equiv
+  (k : localization_map T Q) (j : M ≃* P) (H : j.to_monoid_hom.map S = T) :
+  N ≃* Q :=
+f.mul_equiv_of_localizations $ k.mul_equiv_comap H
+
+-- Localization at a submonoid generated by one element
+
+variables (x : M) (N)
+
+/-- Given `x : M`, the type of monoid homs satisfying the predicate characterizing the
+    localization of `M` at the submonoid generated by `x`. -/
+@[reducible] def away_map := localization_map (submonoid.powers x) N
```

Comment:
In the old localizations code, `away x` was a construction of a localization. Do we already have that here? If so, we need to show that the the localization map to `away x` is an `away_map`.
I'm not so sure if `away_map` should really be a new type... depends on how much we use it.

Reply:
That's all coming in the next PR. (Maybe the file could do with reordering). Hm, I agree ```away_map``` doesn't need to be a new type.

Reply:
Hmmm, I would postpone the `away` stuff for the next PR in that case.
That makes this one really easy and fast to get merged.
And it would make it easier to see the overall picture for `away`.

Reply:
Sure, that sounds sensible.